### PR TITLE
Replace CheckCxxLinkerFlag module with CMake one

### PR DESCRIPTION
**Pull request by wthrowe**
Our module is broken, at least in that it calls charmc.  This requires cmake 3.18.

### Upgrade instructions

The minimum version of CMake has been increased to 3.18.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Review by nilsdeppe — DISMISSED** [review]

**Comment by wthrowe** [comment]
Right, the containers need updating before CI will pass.  I think I've updated the definitions correctly in this PR.  Can someone do the rebuild?

**Comment by nilsdeppe** [comment]
I can try build the container at 12:30pm Pacific. I'll report back if I run into issues, then we can coordinate push to DockerHub and merging this :)

File: cmake/SpectreTargetSources.cmake [hunk 1]
```diff
@@ -26,37 +26,11 @@
...
-    list(APPEND _SOURCE_FILES ${SOURCE_FILE})
-  endforeach()
-  target_sources(${TARGET} ${_SOURCE_FILES})
+  # New behavior is available, so just forward to it by ensuring
+  # that we have the policy set to request the new behavior, but
+  # don't change the policy setting for the calling scope
+  cmake_policy(PUSH)
+  cmake_policy(SET CMP0076 NEW)
+  target_sources(${TARGET} ${ARGN})
+  cmake_policy(POP)
```
Comment (nilsvu) [inline cmake/SpectreTargetSources.cmake]: we don't need this anymore I think. The "new" policy (now always available and set to "new") is to convert relative to absolute paths. So just call `target_sources` and nothing else.

Please update the function docs.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: cmake/SpectreTargetSources.cmake (hunk 1 above)
Comment (wthrowe) [reply cmake/SpectreTargetSources.cmake]: When did the behavior change?  I don't see anything referencing versions between 3.13 and 3.18 in the CMake docs.

Comment (nilsvu) [reply cmake/SpectreTargetSources.cmake]: Changed in 3.13: https://cmake.org/cmake/help/latest/policy/CMP0076.html

Comment (nilsvu) [reply cmake/SpectreTargetSources.cmake]: This means the policy is set to "new" automatically for projects with a min cmake version >= 3.13

**Comment by nilsdeppe** [comment]
I've got the dev and ci environments built. I can push them whenever we want to test this PR on that and then merge.

File: cmake/SpectreTargetSources.cmake (hunk 1 above)
Comment (wthrowe) [reply cmake/SpectreTargetSources.cmake]: Ah, I see.  It's a nonlocal thing.  I'll update it.

**Review by nilsvu — APPROVED** [review]

**Comment by nilsvu** [comment]
@nilsdeppe Do you want to just push the container, then rerun CI, then merge?

**Comment by nilsvu** [comment]
Note that you could do this without disrupting other PRs by pushing an image that still has the old CMake installed (just copy the part of the dockerfile and build). Then the old CMake would get removed from the container next time we build it with the merged version of the dockerfile. Up to you.

**Comment by nilsdeppe** [comment]
Okay, pushed and restart CI

**Pull request merged**